src/main/telemetry/sim.c

```diff
@@ -0,0 +1,379 @@
+        atCommandStatus = SIM_AT_ERROR;
+        if (!simWaitAfterResponse) {
+            sim_t_stateChange = millis() + SIM_AT_COMMAND_DELAY_MIN_MS;
+        }
+        return;
+    } else if (responseCode == SIM_RESPONSE_CODE_RING) {
+        // RING
+    } else if (responseCode == SIM_RESPONSE_CODE_CSQ) {
+        // +CSQ: 26,0
+        simRssi = fastA2I((char*)&simResponse[6]);
+    } else if (responseCode == SIM_RESPONSE_CODE_CLIP) {
+        // we always get this after a RING when a call is incoming
+        // +CLIP: "+3581234567"
+        readOriginatingNumber(&simResponse[8]);
+        if (checkGroundStationNumber(&simResponse[8])) {
+            requestSendSMS();
+        }
+    } else if (responseCode == SIM_RESPONSE_CODE_CMT) {
+        // +CMT: <oa>,[<alpha>],<scts>[,<tooa>,<fo>,<pid>,<dcs>,<sca>,<tosca>,<length>]<CR><LF><data>
+        // +CMT: "+3581234567","","19/02/12,14:57:24+08"
+        readOriginatingNumber(&simResponse[7]);
+        if (checkGroundStationNumber(&simResponse[7])) {
+            readingSMS = true; // next simResponse line will be SMS content
+        } else {
+            // skip SMS content
+            while (serialRxBytesWaiting(simPort)) {
+                if (serialRead(simPort) == '\n') return;
+            }
+        }
+    }
+}
+
+void readOriginatingNumber(uint8_t* rv)
+{
+    int i;
+    uint8_t* gsn = telemetryConfigMutable()->simGroundStationNumber;
+    if (gsn[0] != '\0')
+        return;
+    for (i = 0; i < 15 && rv[i] != '\"'; i++)
+         gsn[i] = rv[i];
+    gsn[i] = '\0';
+}
+
+void readSMS()
+{
+    if (sl_strcasecmp((char*)simResponse, SIM_SMS_COMMAND_TRANSMISSION) == 0) {
+        telemetryConfigMutable()->simTransmissionInterval *= -1;
+        return;
+    } else if (sl_strcasecmp((char*)simResponse, SIM_SMS_COMMAND_RTH) == 0) {
+        if (!posControl.flags.forcedRTHActivated) {
+            activateForcedRTH();
+        } else {
+            abortForcedRTH();
+        }
+    }
+    requestSendSMS();
+}
+
+void detectAccEvents()
+{
+    uint32_t now = millis();
+//    float acceleration = sqrtf(vectorNormSquared(acc.accADCf));
+    uint32_t accSq = sq(imuMeasuredAccelBF.x) + sq(imuMeasuredAccelBF.y) + sq(imuMeasuredAccelBF.z);
+
+    if (telemetryConfig()->accEventThresholdHigh > 0 && accSq > sq(telemetryConfig()->accEventThresholdHigh))
+        accEvent = ACC_EVENT_HIGH;
+    else if (accSq < sq(telemetryConfig()->accEventThresholdLow))
+        accEvent = ACC_EVENT_LOW;
+    else if (telemetryConfig()->accEventThresholdNegX > 0 && imuMeasuredAccelBF.x < -telemetryConfig()->accEventThresholdNegX)
+        accEvent = ACC_EVENT_NEG_X;
+    else
+        return;
+
+    t_accEventDetected = now;
+    if (now - t_accEventMessageSent > 5000) {
+        requestSendSMS();
+        t_accEventMessageSent = now;
+    }
+}
+
+void detectFailsafe()
+{
+    if (!failsafeMessageSent && FLIGHT_MODE(FAILSAFE_MODE) && ARMING_FLAG(ARMED)) {
+        failsafeMessageSent = true;        
```

Comment:
Maybe:
   if (!failsafeMessageSent && FLIGHT_MODE(FAILSAFE_MODE) && ARMING_FLAG(WAS_EVER_ARMED)) {
Multicopter can disarm after landing.
There are also known cases of disarming plane in flight.
So failsafe without arming is passible and should be perceived as an un-normal situation.

Reply:
Ok this is changed now.